python/ray/tune/suggest/search.py

```diff
@@ -0,0 +1,101 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+
+class SearchAlgorithm(object):
+    """SearchAlgorithm exposes an event handler API for hyperparameter search.
+
+    Unlike TrialSchedulers, SearchAlgorithms will not have the ability
+    to modify the execution (i.e., stop and pause trials).
+
+    To track suggestions and their corresponding evaluations, the method
+    `try_suggest` will need to generate a trial_id. This trial_id will
+    be used in subsequent notifications.
+
+    Trials added manually (i.e., via the Client API) will also notify
+    this class upon new events, so custom search algorithms may want to
+    maintain a list of trials ID generated from this class.
+
+    Attributes:
+        NOT_READY (str): Status string for `try_suggest` if SearchAlgorithm
+            currently cannot be queried for parameters (i.e. due to
+            constrained concurrency).
+
+    Example:
+        >>> suggester = SearchAlgorithm()
+        >>> new_parameters, trial_id = suggester.try_suggest()
+        >>> suggester.on_trial_complete(trial_id, result)
+        >>> better_parameters, trial_id2 = suggester.try_suggest()
+    """
+    NOT_READY = "NOT_READY"
+
+    def try_suggest(self):
```

Comment:
Thanks for your comments! 

RE Passing in a search space/custom search space class: You can take a look at tune/suggest/hyperopt.py. The custom search space there is passed in through the init; basically, this particular interface does not dictate a signature.

RE variant generator: OK; this class will then hold a generator that will create trial objects, like `SearchAlgorithm.next_trial`. It will look something like:

```
class SearchAlgorithm:
    def trial_generator(self):
        for resolved_vars in generate_variants:
            config = ...
            suggested = self._suggest() # try_suggest in current API
            if suggested:
                config.update(suggested)
                yield Trial(config)
            else:
                loop ...
                    yield None

    def next_trial(self):
        return next(self.generator)
```
For the most part, `_suggest` will be enough for users that want to implement an algorithm. For more advanced users like yourself, you can just override `trial_generator`.

Will this fit your use case?